### ISSUE TITLE
[JDBC] DuckDB JDBC driver SIGSEGV the JVM since 0.9.0

### ISSUE THREAD

**Issue by loicmathieu** [duckdb/duckdb-java#14]
### What happens?

Since version 0.9.0, using the DuckDB JDBC driver in a Java application makes the application crash with a SIGSEGV.
The Java version is 17.0.5 (tested also on 17.0.8.1).

There is first a Java exception

```
java.sql.SQLException: random_device could not be read
	at org.duckdb.DuckDBNative.duckdb_jdbc_startup(Native Method)
	at org.duckdb.DuckDBConnection.newConnection(DuckDBConnection.java:48)
	at org.duckdb.DuckDBDriver.connect(DuckDBDriver.java:38)
	at java.sql/java.sql.DriverManager.getConnection(DriverManager.java:681)
	at java.sql/java.sql.DriverManager.getConnection(DriverManager.java:190)
	at io.kestra.plugin.jdbc.JdbcConnectionInterface.connection(JdbcConnectionInterface.java:63)
	at io.kestra.plugin.jdbc.AbstractJdbcQuery.run(AbstractJdbcQuery.java:77)
	at io.kestra.plugin.jdbc.duckdb.Query.run(Query.java:148)
	at io.kestra.plugin.jdbc.duckdb.Query.run(Query.java:31)
	at io.kestra.core.runners.Worker$WorkerThread.run(Worker.java:674) 	
```

Then a JVM crash

```
#
# A fatal error has been detected by the Java Runtime Environment:
#
#  SIGSEGV (0xb) at pc=0x00007f523603fd60, pid=37746, tid=39346
#
# JRE version: OpenJDK Runtime Environment Temurin-17.0.5+8 (17.0.5+8) (build 17.0.5+8)
# Java VM: OpenJDK 64-Bit Server VM Temurin-17.0.5+8 (17.0.5+8, mixed mode, sharing, tiered, compressed oops, compressed class ptrs, g1 gc, linux-amd64)
# Problematic frame:
# C  0x00007f523603fd60
#
# Core dump will be written. Default location: Core dumps may be processed with "/usr/share/apport/apport -p%p -s%s -c%c -d%d -P%P -u%u -g%g -- %E" (or dumping to <redacted>)
#
# An error report file with more information is saved as:
# <redacted>
#
# If you would like to submit a bug report, please visit:
#   https://github.com/adoptium/adoptium-support/issues
# The crash happened outside the Java Virtual Machine in native code.
# See problematic frame for where to report the bug.
#
```

It works well with 0.8.0.

### To Reproduce

Here is the SQL query:

```
      INSTALL httpfs;
      SELECT Title, max("Days In Top 10") 
      from (SELECT * FROM read_parquet('s3://duckdb-md-dataset-121/netflix_daily_top_10.parquet'))
      where Type='Movie'
      GROUP BY Title
      ORDER BY max("Days In Top 10") desc
      limit 5;
```

The code uses the standard Java JDBC API (Connection & Statement), but it is not easily extracted as it runs via [Kestra DuckDB plugin](https://github.com/kestra-io/plugin-jdbc/tree/master/plugin-jdbc-duckdb).

### OS:

Ubuntu 23.04

### DuckDB Version:

0.9.0

### DuckDB Client:

Java JDBC

### Full Name:

Loïc Mathieu

### Affiliation:

Kestra

### Have you tried this on the latest `main` branch?

I have tested with a release build (and could not test with a main build)

### Have you tried the steps to reproduce? Do they include all relevant data and configuration? Does the issue you report still appear there?

- [X] Yes, I have

**Comment by armetiz** [comment]
I tried to reproduce the error to help resolve this issue, but It works.

I create a Dockerfile using latest version of `eclipse-temurin`.
Create a Java application that fetch a remote Parquet.

Display the SQL results : ✅

Both 0.9.0 and 0.10.3 are working.

Here the gist with all used files : https://gist.github.com/armetiz/e4ffd81189eb334c5acdf3e9e9796940

Can I try something else to reproduce the problem and hope for a solution?

Regards

**Comment by loicmathieu** [comment]
@armetiz on Kestra, this issue only occurs if the rocksdb native library is loaded before the duckdb native library, this happens in Kestra EE.

**Comment by armetiz** [comment]
Hi @Mause I tried to reproduce [your Maven](https://github.com/Mause/duckdb_rocksdb_crash/blob/main/src/test/java/com/mycompany/app/AppTest.java) configuration within a Docker container.

But as you can see, I could not reproduce the error : https://github.com/armetiz/dockerfile-maven-duckdb-rockdbs

**Comment by elefeint** [comment]
The reproduction of this requires building DuckDB on `manylinux2014` but running the Java application on a modern system. Docker file reproducing the issue with a debug version of DuckDB: [Dockerfile.txt](https://github.com/user-attachments/files/16563761/Dockerfile.txt)

Output:
```
0.256 *** BEFORE LOADING ROCKSDB ***                                                                                                                          
0.363 *** AFTER LOADING ROCKSDB ***                                                                                                                           
2.095 #                                                                                                                                                       
2.095 # A fatal error has been detected by the Java Runtime Environment:                                                                                      
2.095 #
2.095 #  SIGSEGV (0xb) at pc=0x00007ee3aca595cc, pid=7, tid=8
2.095 #
2.095 # JRE version: OpenJDK Runtime Environment Temurin-17.0.12+7 (17.0.12+7) (build 17.0.12+7)
2.095 # Java VM: OpenJDK 64-Bit Server VM Temurin-17.0.12+7 (17.0.12+7, mixed mode, sharing, tiered, compressed oops, compressed class ptrs, g1 gc, linux-amd64)
2.095 # Problematic frame:
2.095 # C  [libduckdb_java10883698283861250744.so+0x2cec5cc]  duckdb::Vector::GetVectorType() const+0xc
2.096 #
2.096 # Core dump will be written. Default location: //core.7
2.096 #
2.096 # An error report file with more information is saved as:
2.096 # //hs_err_pid7.log
2.249 #
2.249 # If you would like to submit a bug report, please visit:
2.249 #   https://github.com/adoptium/adoptium-support/issues
2.249 # The crash happened outside the Java Virtual Machine in native code.
2.249 # See problematic frame for where to report the bug.
2.249 #
2.460 Aborted (core dumped)
```

**Comment by Ben8t** [comment]
@Mause any update on this ?

**Comment by Ben8t** [comment]
Comments from @elefeint on Slack
> DuckDB labs has to balance supporting older systems with the problems that the older system libraries sometimes cause. We'll follow up to see if there are any plans on DuckDB side to move up from manylinux2014 for their builds, but I am curious -- from the Kestra side, is it necessary to include all the drivers in a Kestra installation? This problem only manifests when both RocksDB and DuckDB are both on the classpath, and the population of users who need both simultaneously is likely small.
The other way you could unblock Kestra users is by building DuckDB from source on a newer base image.

@loicmathieu does it help in someway ? Or best option is to wait for followup on "plans on DuckDB side to move up from manylinux2014 for their builds" ?

**Comment by loicmathieu** [comment]
I think I told them in their Slack, but we use Kafka Stream that uses RocksDB so we don't have the choice here.
We will need to wait but this issue is one year old now ...

**Comment by anna-geller** [comment]
> This problem only manifests when both RocksDB and DuckDB are both on the classpath, and the population of users who need both simultaneously is likely small.

Seeing the number of upvotes, comments and the number of kestra users asking about it on our Slack, the population of simultaneous RocksDB and DuckDB users is not small. The issue has been open for a while now; @Mause @elefeint, could someone give specific guidance on how to proceed or fix this on the DuckDB side? This is a large hindrance for many users and we are blocked.

**Comment by shoffmeister** [comment]
I am getting this, too, and for me it is happening in a CI build container, with Kafka Streams + RocksDB + DuckDB.

**Comment by Mause** [comment]
> > This problem only manifests when both RocksDB and DuckDB are both on the classpath, and the population of users who need both simultaneously is likely small.
> 
> Seeing the number of upvotes, comments and the number of kestra users asking about it on our Slack, the population of simultaneous RocksDB and DuckDB users doesn't seem to be that small. The issue has been open for a while now; @Mause @elefeint, could someone give specific guidance on how to proceed or fix this on the DuckDB side? This is a large hindrance for many users and we are blocked. 

Can I ask for a link to the RocksDB issue?

**Comment by anna-geller** [comment]
Sorry if it was confusing: I meant Kestra users who use RocksDB indirectly. I'm not aware of any RocksDB issue about this

**Comment by Mause** [comment]
Could you please raise one then? I'd be curious to hear what they have to say on this issue, given it seems to be something they're doing to cause this

**Comment by shoffmeister** [comment]
FWIW, I am using "plain" Spring Boot Cloud Kafka Streams (which obviously includes RocksDB), and I do see failure in running unit tests in a specific Maven CI container (from Docker Hub). 

Note: I only use DuckDB in the unit tests, so RocksDB definitely loads first. There is nothing Kestra here on my end. I do acknowledge that Kestra may have material commercial interest in getting that fixed. ;)

Since ELF is late-binding by default, and looking at the LD_PRELOAD magic that's being mentioned, I suspect that the (plain) Kafka Streams JAR (https://central.sonatype.com/artifact/org.apache.kafka/kafka-streams/versions) bundles a binary copy of RocksDB (which is C++, see https://github.com/facebook/rocksdb/blob/main/CMakeLists.txt) which has been built in a way, and with flags, that is ... "unexpected"?

After all, all the LD_PRELOAD magic does is _early-resolve_ the symbols in libstdc++, and _to symbols from that very same ELF binary_.

I'll try to amend https://github.com/Mause/duckdb_rocksdb_crash with a containerized build script that allows for global reproducibility. I believe this is critical, as on my personal development environment (which is Arch Linux), all is - or appears to be! - well.

**Comment by elefeint** [comment]
@shoffmeister you can use the Dockerfile in https://github.com/duckdb/duckdb-java/issues/14#issuecomment-2278211096 to reproduce

**Comment by shoffmeister** [comment]
Using the baseline that @Mause created in https://github.com/Mause/duckdb_rocksdb_crash, this simple script will reproduce:

```
#!/usr/bin/env bash

# https://hub.docker.com/_/maven
# https://github.com/carlossg/docker-maven/blob/8cfe24baffa5b250f7bb2d31ce233fc28f3c4f20/eclipse-temurin-17/Dockerfile
IMAGE=maven:3-eclipse-temurin-17
IMAGE=maven@sha256:cf1bca11a285e887efebe851d8e55e4defa326b7ca29a68920f1c9dccc5dad4f

docker run -it --rm --name duckdb-rocksdb_crash \
    -u "$(id -u):$(id -g)" \
    -v "$(pwd)":/usr/src/build:rw \
    -v "$HOME/.m2":/usr/src/.m2:rw \
    -e "MAVEN_OPTS=-Dmaven.repo.local=/usr/src/.m2/repository" \
    -w /usr/src/build \
    ${IMAGE} \
    mvn --offline clean verify
```

Assumptions:
* your Maven `settings.xml` does not exist (or is fully usable inside a container)
* `mvn clean package` once before running inside the container

Running that script then yields
```
SIGSEGV (0xb) at pc=0x00007d0bb4fc00e0, pid=1, tid=26
```

The image used above is a very recent Ubuntu 24.04.01, directly from Docker Hub, with a very recent Maven, and Java 17. Since that is a moving target, the digest nails the exact image identity for easier (future) reproducibility.

Adding 
```
    -e "LD_PRELOAD=/lib/x86_64-linux-gnu/libstdc++.so.6" \
```
to the mix removes the crash. Alas, adding `LD_DEBUG=all` hangs and does not yield useful output beyond

```
/usr/bin/java: error: symbol lookup error: undefined symbol: JNI_OnLoad_rocksdbjni-linux64 (fatal)
```
both inside the container and on the host. I suspect that this is due to the way the Java reproducer has been done, and that forking gets into the way.

While I used https://github.com/Mause/duckdb_rocksdb_crash to reproduce, this most likely mirrors my local setup where RocksDB gets pulled by Kafka Streams (which is pulled by Spring Cloud Kafka)

To me it is somewhat surprising that my local development environment is totally fine (Arch Linux as a rolling always-up-to-date distribution), but that the rather modern Maven build container shows the problem. 

This would suggest that the way the _runtime_ distribution's user land has been built also plays a role.

**Comment by shoffmeister** [comment]
There is a different failure mode when building the reproducer as a fat JAR _application_ (outside of the container) and then running that inside the container: the process simply hangs.

With the above it is also clear, IMHO, that the distribution on which this is run does matter.

**Comment by shoffmeister** [comment]
The answer might be hiding in the (differential) output of
```
LD_DEBUG=symbols LD_DEBUG_OUTPUT=debug.log java -jar target/my-app-1.0-SNAPSHOT-jar-with-dependencies.jar
```
where the JAR is the fat JAR, and this is being run inside two different Java-enabled containers with different user libraries, where one container shows the problem, the other not.

**Comment by shoffmeister** [comment]
The content of https://github.com/tensorflow/tensorflow/issues/61741 is intriguing.

My _initial_ reproducer in a large Kafka Streams application actually (also) shows
```
java.sql.SQLException: Invalid Error: random_device could not be read: Bad file descriptor
	at org.duckdb.DuckDBNative.duckdb_jdbc_startup(Native Method) ~[duckdb_jdbc-1.1.0.jar:na]
```
which I believe has the same root cause as this crash here.

**Comment by loicmathieu** [comment]
The issue at RocksDB side: https://github.com/facebook/rocksdb/issues/13092
Hope they will be able to help us.

**Comment by shoffmeister** [comment]
There is interesting analysis at https://github.com/pytorch/pytorch/issues/102360 in the comments, specifically on the subject of `USE_POSIX_FILE_IO` and `_GLIBCXX_USE_CXX11_ABI` and binding of imports.

They are fighting with what I gather to be a very similar problem; the root cause there apparently is "cross-talk" of C++ std::random functions from dynamically loaded binaries. They also preload the system libstdc++ which is the obvious solution in their case.

If I interpret
```
        11:	symbol=_ZNSt13random_device7_M_initERKSs;  lookup in file=java [0]
        11:	symbol=_ZNSt13random_device7_M_initERKSs;  lookup in file=/opt/java/openjdk/bin/../lib/libjli.so [0]
        11:	symbol=_ZNSt13random_device7_M_initERKSs;  lookup in file=/lib/x86_64-linux-gnu/libpthread.so.0 [0]
        11:	symbol=_ZNSt13random_device7_M_initERKSs;  lookup in file=/lib/x86_64-linux-gnu/libdl.so.2 [0]
        11:	symbol=_ZNSt13random_device7_M_initERKSs;  lookup in file=/lib/x86_64-linux-gnu/libc.so.6 [0]
        11:	symbol=_ZNSt13random_device7_M_initERKSs;  lookup in file=/lib64/ld-linux-x86-64.so.2 [0]
        11:	symbol=_ZNSt13random_device7_M_initERKSs;  lookup in file=/opt/java/openjdk/lib/server/libjvm.so [0]
        11:	symbol=_ZNSt13random_device7_M_initERKSs;  lookup in file=/lib/x86_64-linux-gnu/librt.so.1 [0]
        11:	symbol=_ZNSt13random_device7_M_initERKSs;  lookup in file=/lib/x86_64-linux-gnu/libm.so.6 [0]
        11:	symbol=_ZNSt13random_device7_M_initERKSs;  lookup in file=/tmp/libduckdb_java14682458476252195639.so [0]
        11:	symbol=_ZNSt13random_device7_M_initERKSs;  lookup in file=/lib/x86_64-linux-gnu/libdl.so.2 [0]
        11:	symbol=_ZNSt13random_device7_M_initERKSs;  lookup in file=/lib/x86_64-linux-gnu/libstdc++.so.6 [0]
        11:	symbol=_ZNSt13random_device7_M_initERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEE;  lookup in file=java [0]
        11:	symbol=_ZNSt13random_device7_M_initERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEE;  lookup in file=/opt/java/openjdk/bin/../lib/libjli.so [0]
        11:	symbol=_ZNSt13random_device7_M_initERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEE;  lookup in file=/lib/x86_64-linux-gnu/libpthread.so.0 [0]
        11:	symbol=_ZNSt13random_device7_M_initERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEE;  lookup in file=/lib/x86_64-linux-gnu/libdl.so.2 [0]
        11:	symbol=_ZNSt13random_device7_M_initERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEE;  lookup in file=/lib/x86_64-linux-gnu/libc.so.6 [0]
        11:	symbol=_ZNSt13random_device7_M_initERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEE;  lookup in file=/lib64/ld-linux-x86-64.so.2 [0]
        11:	symbol=_ZNSt13random_device7_M_initERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEE;  lookup in file=/opt/java/openjdk/lib/server/libjvm.so [0]
        11:	symbol=_ZNSt13random_device7_M_initERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEE;  lookup in file=/lib/x86_64-linux-gnu/librt.so.1 [0]
        11:	symbol=_ZNSt13random_device7_M_initERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEE;  lookup in file=/lib/x86_64-linux-gnu/libm.so.6 [0]
        11:	symbol=_ZNSt13random_device7_M_initERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEE;  lookup in file=/tmp/librocksdbjni1494878192266256145.so [0]
```
right, from the LD_DEBUG above, in the affected container, and then add
```
nm --demangle /tmp/librocksdbjni1494878192266256145.so | grep random_device
```
getting 
```
0000000000a7b080 T std::random_device::_M_init_pretr1(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)
0000000000a80f60 T std::random_device::_M_init_pretr1(std::string const&)
0000000000a7b280 T std::random_device::_M_getval_pretr1()
0000000000a7af90 T std::random_device::_M_fini()
0000000000a7aee0 T std::random_device::_M_init(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)
0000000000a80eb0 T std::random_device::_M_init(std::string const&)
0000000000a7afb0 T std::random_device::_M_getval()
```

This means that RocksDB contains these symbols, and symbol resolution eventually hits that.

The difference between my local Arch Linux machine and the container in the libstdc++ library is

container:
```
  4840: 00000000000e9910  1024 FUNC    GLOBAL DEFAULT   15 _ZNSt13random_device7_M_initERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEE@@GLIBCXX_3.4.21
```

arch:
```
  4841: 00000000000e0950  1045 FUNC    GLOBAL DEFAULT   12 _ZNSt13random_device7_M_initERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEE@@GLIBCXX_3.4.21
   527: 000000000009c01a   122 FUNC    LOCAL  DEFAULT   12 _ZNSt13random_device7_M_initERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEE.cold
  5799: 00000000000e0950  1045 FUNC    GLOBAL DEFAULT   12 _ZNSt13random_device7_M_initERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEE
```
so my Arch provides `_ZNSt13random_device7_M_initERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEE` which RocksDB likes to work with.

In the Ubuntu container I reproduce with, libstdc++ does not have the _general_ export, so DuckDB seems to be binding to the equivalent export in RocksDB upon the attempt to init the random subsystem. And that ... is not good.

From the looks of it, the RocksDB shared object should not contain those exports, i.e.
```
readelf --wide --syms /tmp/librocksdbjni1494878192266256145.so | grep random_device
```
should return nothing.

Note that `/tmp/librocksdbjni1494878192266256145.so` is what the JVM extracts into /tmp from the RocksDB JAR as the JNI library.

**Comment by loicmathieu** [comment]
@shoffmeister maybe you can give this information in the issue I just raised at RockDB side: https://github.com/facebook/rocksdb/issues/13092

**Comment by shoffmeister** [comment]
Added some context on the RocksDB repository asking why
```
unzip -o rocksdbjni-9.6.1.jar librocksdbjni-linux64.so -d . &&  nm --demangle ./librocksdbjni-linux64.so | cut -c 18- | grep 'T std::' | sort
```
shows so many symbols exported _from_ RocksDB in the `std::` namespace.

**Comment by shoffmeister** [comment]
FWIW, I just took a look at the DuckDB JNI
```
nm --demangle ./libduckdb_java15504998828832578825.so | cut -c 18- | grep 'T '
```
or
```
readelf --syms --wide --demangle ./libduckdb_java15504998828832578825.so
```
and it would seem as if the DSO also _exports_ a massive amount of symbols beyond need. That does not isolate well, so DuckDB also seems to be doing what RocksDB is doing. cc'ing @carlopi because he seems to be the resident "having fun with operating system library linkage" person ;)

Example:
```
FUNC    GLOBAL DEFAULT   11 icu_66::double_conversion::Bignum::SubtractBignum(icu_66::double_conversion::Bignum const&)
```

As mentioned for RocksDB, I would only expect those exported:
```
readelf --syms --wide --demangle ./libduckdb_java15504998828832578825.so | grep Java_org_duckdb
```
because anything else pollutes the ELF symbol namespace, interferes with (late-binding) symbol resolution, and can have side-effects. A JNI should not have symbol side-effects on being loaded into a native operating system process, IMHO.

**Comment by pdet** [comment]
Hi guys,

I've pushed a fix for this issue in:
https://github.com/duckdb/duckdb/pull/15540

I've verified it worked with the docker file provided by @elefeint here: https://github.com/duckdb/duckdb-java/issues/14#issuecomment-2278211096

But it would be interesting if others could also validate that this fixes their issues on different platforms as well :-)

**Comment by FelixGV** [comment]
Hi folks!

It's great to see that this issue has been fixed!

I am wondering: what is the process for the DuckDB project to cut a new stable release on Maven Central? The last one, 1.1.3, was published on 2024-11-05, and depending on SNAPSHOT dependencies complicates things...

I'm looking forward to trying out DuckDB!

Cheers!

**Comment by FelixGV** [comment]
BTW, I've confirmed via a SNAPSHOT dependency that the fix does work for us. Thanks! Looking forward to an official release containing this!

**Comment by pdet** [comment]
> BTW, I've confirmed via a SNAPSHOT dependency that the fix does work for us. Thanks! Looking forward to an official release containing this!

It's scheduled to Jan 27 :-)
https://duckdb.org/docs/dev/release_calendar.html

**Comment by shivanshu-bootnext** [comment]
I am getting this error while running the duck db service inside the docker container. I am working on the in memory implementation.

Failed to initialize DuckDB: No error information
java.sql.SQLException: No error information
	at org.duckdb.DuckDBNative.duckdb_jdbc_startup(Native Method)
	at org.duckdb.DuckDBConnection.newConnection(DuckDBConnection.java:48)
	at org.duckdb.DuckDBDriver.connect(DuckDBDriver.java:38)
	at java.sql/java.sql.DriverManager.getConnection(DriverManager.java:683)
	at java.sql/java.sql.DriverManager.getConnection(DriverManager.java:253)
	at com.enttribe.duckdb.service.impl.DuckDbServiceImpl.init(DuckDbServiceImpl.java:22)

**Comment by loicmathieu** [comment]
@shivanshu-bootnext this seems to be a different issue.
Please open a new issue with a reproducer.
